src/GraphQL/Resolver/LocationGuesser/TreeRootLocationFilter.php

```diff
@@ -0,0 +1,123 @@
+    {
+        $this->locationService = $locationService;
+        $this->configResolver = $configResolver;
+        $this->urlAliasService = $urlAliasService;
+    }
+
+    public function filter(Content $content, LocationList $locationList): void
+    {
+        foreach ($locationList->getLocations() as $location) {
+            if (!$this->locationIsInTreeRoot($location) && !$this->locationPrefixIsExcluded($location)) {
+                $locationList->filter($location);
+            }
+        }
+    }
+
+    /**
+     * Checks if a location is valid in regards to the tree root setting.
+     *
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     *
+     * @return bool
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    private function locationIsInTreeRoot(Location $location): bool
+    {
+        static $rootLocations = [];
+
+        $treeRootLocationId = $this->configResolver->getParameter('content.tree_root.location_id');
+        if (!isset($rootLocations[$treeRootLocationId])) {
+            $rootLocations[$treeRootLocationId] = $this->locationService->loadLocation($treeRootLocationId);
+        }
+
+        return $this->containsRootPath($location->path, $rootLocations[$treeRootLocationId]->path);
+    }
+
+    /**
+     * Tests if the location is excluded from tree root.
+     *
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $candidateLocation
+     *
+     * @return bool
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    private function locationPrefixIsExcluded(Location $candidateLocation): bool
+    {
+        static $excludedLocations = null;
+
+        if ($excludedLocations === null) {
+            $excludedUriPrefixes = $this->configResolver->getParameter('content.tree_root.excluded_uri_prefixes');
+            if (empty($excludedUriPrefixes)) {
+                return false;
+            }
+            foreach ($excludedUriPrefixes as $uri) {
+                $urlAlias = $this->urlAliasService->lookup($uri);
+                if ($urlAlias->type === URLAlias::LOCATION) {
+                    $excludedLocations[] = $this->locationService->loadLocation($urlAlias->destination);
+                }
+            }
+        }
+
+        foreach ($excludedLocations as $excludedLocation) {
+            if ($this->containsRootPath($candidateLocation->path, $excludedLocation->path)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param array $path
+     * @param array $rootPath
+     *
+     * @return bool
+     */
+    private function containsRootPath(array $path, array $rootPath): bool
+    {
+        return array_slice($path, 0, count($rootPath)) === $rootPath;
+    }
```

Comment:
We should have Path VO into PAPI to avoid re-implementing path operation at-hoc...

Reply:
Fine by me.

Reply:
We could even go further, and have an official API for getting the URL alias for a location for a given siteaccess.